activesupport/lib/active_support/actionable_error.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+require "active_support/concern"
+
+module ActiveSupport
+  # Actionable errors let's you define actions to resolve an error.
+  #
+  # To make an error actionable, include the <tt>ActiveSupport::ActionableError</tt>
+  # module and invoke the +action+ class macro to define the action. An action
+  # needs a name and a block to execute.
+  module ActionableError
+    extend Concern
+
+    NonActionable = Class.new(StandardError)
+
+    NoActions = Hash.new do |_, name| # :nodoc:
+      raise NonActionable, "Cannot find action \"#{name}\""
+    end
+
+    included do
+      class_attribute :_actions, default: NoActions.dup
+    end
+
+    def self.actions(error) # :nodoc:
+      case error
+      when ActionableError, -> it { Class === it && it < ActionableError }
+        error._actions
+      else
+        NoActions
+      end
+    end
+
+    def self.dispatch(error, name) # :nodoc:
+      actions(error.is_a?(String) ? error.constantize : error)[name].call
+    end
```

Comment:
I get the NoActions, but is it really clearer than something like this?

```ruby
def actions(error)
  if error < ActionableError
    error._actions
  else
    {}
  end
end

def dispatch(error, name)
  actions(error).fetch(name).call
rescue KeyError
  raise NonActionable, "Cannot find action \"#{name}\""
end
```

Reply:
Fair enough will reduce the magic furthermore. 👍